### ISSUE TITLE
Feature: Top pathway hit

### ISSUE THREAD

**Issue by jvwong** [PathwayCommons/app-ui#1425]
**Goal**
Feature a the top pathway search hit and include a visual preview (SBGN) when a query references at least one of the molecular participants. 

**Primary use case**
Referral from public entity databases. In these cases, the query will be in the form of a (a) UniProt accession (referrer: UniProt) and/or b) HGNC symbol (referrer: GeneCards). The response should contain information necessary to populate a Feature Component for the pathway: name, source, #participants, and a URL pointing to an SBGN preview]

**Restrictions & Limitations**
Main concern is slowing down the "time to interactive" of the search, which degrades the user experience. Several potential  bottlenecks:
  - Generating an SBGN snapshot on-the-fly
  - Identifying entity IDs from the query
  - Retrieving a pathway's participants (UnificationXref db, id).

**Edge cases**
Integrating with case where Biofactoid is contained in a Feature view (i.e. triggered when top search hit source is Biofactoid)

**Notes**
To reduce bottlenecks, might consider

a. Rendering and caching SBGN previews for all pathways instread of on-the-fly, which might also be useful in Google indexing
  - https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1010635
b. Have cPath2 index and return UnificationXrefs to bypass need to retrieve these on-the-fly information separately

**Comment by jvwong** [comment]
### Update

**Goal**
Feature a the top pathway search hit and include a visual preview (SBGN) so that users are aware of what the pathway section links will hold.

**Primary use case**
Any user that visits the search page.

**Mockup**

![summary](https://user-images.githubusercontent.com/4706307/204637850-614b4981-b56e-41e0-97e3-a39a3f342fd6.png)

**Details**

Benefits:
- Generic to all users of the search, whether referred (UniProt, GeneCards) or arriving via Google
- Does not change the experience for majority of users who are mainly interested in interactions app 
- Preview is placed in the Pathway section, indicating that other links will provide the same info (SBGN)
- No complex mapping of query to pathway previewed
- User existing components (e.g. SBGN pathway viewer)

Risks:
- The top hit may or may not be relevant (i.e. we're not checking that a database ID query is in the pathways)
- Generating the SBGN visualization may be slow
- SBGN network can be enormous, complex

**Comment by cannin** [comment]
Python code to generate static images of PC pathways using the Syblars (https://pubmed.ncbi.nlm.nih.gov/36374853/ from Ugur Dogrusoz)

## Code 
https://gist.github.com/cannin/7e35f3fae274370bd0a70c7b1840c743

## General Workflow
1. Use a PC GMT file to get pathway URIs
2. Call Pathway Commons API to get SBGN
3. Call Syblars to get PNG data content 
4. Save to file 

# Examples
There is an upper limit nodes where the automated layout of SBGN is poor (~100 nodes???)

![http-identifiers-org-kegg-pathway-hsa00601](https://github.com/PathwayCommons/app-ui/assets/790284/146d22d4-946a-423a-9dbf-1c985887001c)
![http-identifiers-org-kegg-pathway-hsa00860](https://github.com/PathwayCommons/app-ui/assets/790284/1e8439d0-d2ff-4cdf-9bd6-15e48bcee38a)
![http-identifiers-org-kegg-pathway-hsa00920](https://github.com/PathwayCommons/app-ui/assets/790284/126f2ea2-f4d4-49ee-bbb2-8161741c45c3)
![http-identifiers-org-kegg-pathway-hsa01100](https://github.com/PathwayCommons/app-ui/assets/790284/4fb96745-aa83-4f0d-ab06-bbc26379382a)

# Indexing of Static Content by Google
Google depends on sitemaps to know what to index. Example from another project with 1700 pages that includes shows minimum to have static images indexed : https://discover.nci.nih.gov/rsconnect/cellminercdb/cell_lines/sitemap.xml Pages were slow to index at first, but everything eventually indexed; images tend to be even slower.

**Comment by jvwong** [comment]
I looked into using the existing services here in `app-ui`, and the only blocking issue in generating image snapshots with cytosnap is the lack of support for styles via [cytoscape-sbgn-stylesheet](https://www.npmjs.com/package/cytoscape-sbgn-stylesheet), as mentioned in this [issue](https://github.com/cytoscape/cytosnap/issues/26). In fact, I think the latter issue was posted in reference to [syblars](https://github.com/iVis-at-Bilkent/syblars) and I suspect they've ended up using a cytosnap fork of some sort.

I was also thinking that also/instead we could pre-cache the pathway JSON data - PC HGNC GMT file says there are 3971 pathways.

**Comment by maxkfranz** [comment]
>  and I suspect they've ended up using a cytosnap fork of some sort.

This looks like it: https://github.com/iVis-at-Bilkent/cytosnap/commits/master

There are a number of new features there, including SVG export.  It would be best to get those merged into the main Cytosnap lib.  The main thing for now is SBGN support.

@ugurdogrusoz, @hasanbalci -- are there any technical reasons why the general stylesheet / SBGN support could not be merged into the main lib?  If not, we should aim to have features like this merged into the main Cytosnap lib so that it's maintained and compatible with the rest of the ecosystem going forward.

**Comment by maxkfranz** [comment]
Great.  Thanks, @hasanbalci

**Comment by jvwong** [comment]
This first pass is running on our [appsbeta.pathwaycommons.org](https://appsbeta.pathwaycommons.org/). There's some good, bad and ugly, depending on the pathway size:

### Good

[Mitochondrial transcription termination | Reactome](https://appsbeta.pathwaycommons.org/search?q=Mitochondrial%20transcription%20termination&type=Pathway)

<img src="https://github.com/PathwayCommons/app-ui/assets/4706307/8e2e9add-a1f8-455f-a685-2732d62c73e0" width="600px" />

### Bad

[TP53 regulates transcription of additional cell cycle genes... Reactome](https://appsbeta.pathwaycommons.org/search?q=TP53%20regulates%20transcription%20of%20additional%20cell%20cycle%20genes%20whose%20exact%20role%20in%20the%20p53%20pathway%20remain%20uncertain&type=Pathway)

<img src="https://github.com/PathwayCommons/app-ui/assets/4706307/1502f80d-2a61-4381-9800-9257df447cf7" width="600px" />

### Ugly

[Direct p53 effectors | NCI-PID](https://appsbeta.pathwaycommons.org/search?q=Direct%20p53%20effectors&type=Pathway)

<img src="https://github.com/PathwayCommons/app-ui/assets/4706307/d1a12eae-b3fc-41b9-9ca1-83506c6256ff"  width="600px" />

**Comment by maxkfranz** [comment]
Improving the network layout generally would at least remove the 'ugly' (hairball) cases, even if they're still 'bad' (too large).  

This could motivate bumping up the priority of updated layout as a student project.  We have lots of new layouts to try that didn't exist when the PC apps were built.

**Comment by jvwong** [comment]
> Improving the network layout generally would at least remove the 'ugly' (hairball) cases, even if they're still 'bad' (too large).
> 
By "bad", I meant from the standpoint of "how is this helpful or attractive to a user" - not so much if text labels are unreadable. So maybe better zooming, panning, and cropping (like google maps).

> This could motivate bumping up the priority of updated layout as a student project. We have lots of new layouts to try that didn't exist when the PC apps were built.

Sounds good.

**Comment by maxkfranz** [comment]
Yeah, the 'bad' is a balance. On the one hand, a zoomed in view is prettier. On the other hand, a full view is more what-you-see-is-what-you-get. 

Either way, better layout would be good.

**Comment by jvwong** [comment]
Just to re-up on my previous comment: https://github.com/PathwayCommons/app-ui/issues/1425#issuecomment-1852076518 

The rendering and layouts are done in syblars: It would be nice (preferred) if we could just do all the rendering/snapping/layouts here.

**Comment by jvwong** [comment]
Just putting here: In regards to how my command line script generates names of images, an outstanding issue (@IgorRodchenkov)

https://github.com/PathwayCommons/app-ui/pull/1443#issuecomment-1999675019
